api/src/services/opportunities_v0_1/search_opportunities.py

```diff
@@ -1,20 +1,117 @@
+    filters: SearchOpportunityFilters | None = Field(default=None)
+
+
+def _add_query_filters(
+    stmt: Select[tuple[Opportunity]], query: str | None
+) -> Select[tuple[Opportunity]]:
+    if query is None or len(query) == 0:
+        return stmt
+
+    # TODO - will implement this in https://github.com/HHS/simpler-grants-gov/issues/1455
+
+    return stmt
+
+
+def _add_filters(
+    stmt: Select[tuple[Opportunity]], filters: SearchOpportunityFilters | None
+) -> Select[tuple[Opportunity]]:
+    if filters is None:
+        return stmt
+
+    # We need to add joins so that the where clauses
+    # can query against the tables that are relevant for these filters
+    # Current + Opportunity Summary are always needed so just add them here
+    stmt = stmt.join(CurrentOpportunitySummary).join(
+        OpportunitySummary,
+        CurrentOpportunitySummary.opportunity_summary_id
+        == OpportunitySummary.opportunity_summary_id,
+    )
+
+    if filters.opportunity_status is not None:
+        one_of_opportunity_statuses = filters.opportunity_status.get("one_of")
+
+        if one_of_opportunity_statuses:
+            stmt = stmt.where(
+                CurrentOpportunitySummary.opportunity_status.in_(one_of_opportunity_statuses)
+            )
+
+    if filters.funding_instrument is not None:
+        stmt = stmt.join(LinkOpportunitySummaryFundingInstrument)
+
+        one_of_funding_instruments = filters.funding_instrument.get("one_of")
+        if one_of_funding_instruments:
+            stmt = stmt.where(
+                LinkOpportunitySummaryFundingInstrument.funding_instrument.in_(
+                    one_of_funding_instruments
+                )
+            )
+
+    if filters.funding_category is not None:
+        stmt = stmt.join(LinkOpportunitySummaryFundingCategory)
+
+        one_of_funding_categories = filters.funding_category.get("one_of")
+        if one_of_funding_categories:
+            stmt = stmt.where(
+                LinkOpportunitySummaryFundingCategory.funding_category.in_(
+                    one_of_funding_categories
+                )
+            )
+
+    if filters.applicant_type is not None:
+        stmt = stmt.join(LinkOpportunitySummaryApplicantType)
+
+        one_of_applicant_types = filters.applicant_type.get("one_of")
+        if one_of_applicant_types:
+            stmt = stmt.where(
+                LinkOpportunitySummaryApplicantType.applicant_type.in_(one_of_applicant_types)
+            )
+
+    if filters.agency is not None:
+        # Note that we filter against the agency code in the opportunity, not in the summary
+        one_of_agencies = filters.agency.get("one_of")
+        if one_of_agencies:
+            # This produces something like:
+            #   ..WHERE ((opportunity.agency ILIKE 'US-ABC%') OR (opportunity.agency ILIKE 'HHS%'))
+            stmt = stmt.where(
+                or_(*[Opportunity.agency.istartswith(agency) for agency in one_of_agencies])
+            )
```

Comment:
It looks like you have an index on agency, so this would do an index scan but not a direct index lookup. Would a direct index lookup on agency be faster? 

Also I was reading that `ilike` does a case insensitive `lower(column_name)` search - is the agency index setup with lowercase to accommodate that?

Reply:
Did some testing (see below). No difference between `like` and `ilike` from what I see. Changing to an exact match was a lot better performance. `ilike` does a sequential scan, and `=` does a Bitmap index scan.

---

A quick check, using the query in the summary, but with only the is_draft, current opportunity_summary, and agency filters (eg. agency bit like `AND((opportunity.agency ILIKE 'US-MMM%') OR(opportunity.agency ILIKE 'HHS%'))`) gives 
```
Limit  (cost=357.66..357.67 rows=1 width=207) (actual time=9.897..9.899 rows=0 loops=1)
  ->  Sort  (cost=357.65..357.66 rows=4 width=207) (actual time=9.896..9.897 rows=0 loops=1)
        Sort Key: opportunity.opportunity_id DESC
        Sort Method: quicksort  Memory: 25kB
        ->  Nested Loop  (cost=1.71..357.61 rows=4 width=207) (actual time=9.855..9.857 rows=0 loops=1)
              Join Filter: (current_opportunity_summary.opportunity_summary_id = link_opportunity_summary_funding_category.opportunity_summary_id)
              ->  Nested Loop  (cost=1.43..356.55 rows=3 width=239) (actual time=9.855..9.856 rows=0 loops=1)
                    Join Filter: (current_opportunity_summary.opportunity_summary_id = link_opportunity_summary_funding_instrument.opportunity_summary_id)
                    ->  Nested Loop  (cost=1.14..355.84 rows=2 width=231) (actual time=9.854..9.855 rows=0 loops=1)
                          ->  Nested Loop Semi Join  (cost=0.86..355.48 rows=1 width=223) (actual time=9.854..9.855 rows=0 loops=1)
                                Join Filter: (opportunity.opportunity_id = current_opportunity_summary_1.opportunity_id)
                                ->  Nested Loop  (cost=0.57..355.15 rows=1 width=231) (actual time=9.854..9.854 rows=0 loops=1)
                                      ->  Nested Loop  (cost=0.29..354.81 rows=1 width=223) (actual time=9.853..9.854 rows=0 loops=1)
                                            ->  Seq Scan on opportunity  (cost=0.00..346.50 rows=1 width=207) (actual time=9.853..9.853 rows=0 loops=1)
"                                                  Filter: ((is_draft IS FALSE) AND ((agency ~~* 'US-MMM%'::text) OR (agency ~~* 'HHS%'::text)))"
                                                  Rows Removed by Filter: 10500
                                            ->  Index Scan using current_opportunity_summary_opportunity_id_idx on current_opportunity_summary  (cost=0.29..8.30 rows=1 width=16) (never executed)
                                                  Index Cond: (opportunity_id = opportunity.opportunity_id)
                                      ->  Index Only Scan using opportunity_summary_pkey on opportunity_summary  (cost=0.29..0.34 rows=1 width=8) (never executed)
                                            Index Cond: (opportunity_summary_id = current_opportunity_summary.opportunity_summary_id)
                                            Heap Fetches: 0
                                ->  Index Only Scan using current_opportunity_summary_opportunity_id_idx on current_opportunity_summary current_opportunity_summary_1  (cost=0.29..0.32 rows=1 width=8) (never executed)
                                      Index Cond: (opportunity_id = current_opportunity_summary.opportunity_id)
                                      Heap Fetches: 0
                          ->  Index Only Scan using link_opportunity_summary_applicant_type_opportunity_sum_fee2 on link_opportunity_summary_applicant_type  (cost=0.29..0.34 rows=2 width=8) (never executed)
                                Index Cond: (opportunity_summary_id = current_opportunity_summary.opportunity_summary_id)
                                Heap Fetches: 0
                    ->  Index Only Scan using link_opportunity_summary_funding_instrument_opportunity_4597 on link_opportunity_summary_funding_instrument  (cost=0.29..0.33 rows=2 width=8) (never executed)
                          Index Cond: (opportunity_summary_id = link_opportunity_summary_applicant_type.opportunity_summary_id)
                          Heap Fetches: 0
              ->  Index Only Scan using link_opportunity_summary_funding_category_opportunity_s_9f79 on link_opportunity_summary_funding_category  (cost=0.29..0.33 rows=2 width=8) (never executed)
                    Index Cond: (opportunity_summary_id = link_opportunity_summary_applicant_type.opportunity_summary_id)
                    Heap Fetches: 0
Planning Time: 27.059 ms
Execution Time: 10.096 ms
```

Changing those queries to `=` rather than `ILIKE` (eg. `AND((opportunity.agency = 'US-MMM') OR(opportunity.agency = 'HHS'))`) gives:
```
Limit  (cost=23.76..23.77 rows=1 width=207) (actual time=0.346..0.348 rows=0 loops=1)
  ->  Sort  (cost=23.75..23.76 rows=4 width=207) (actual time=0.345..0.346 rows=0 loops=1)
        Sort Key: opportunity.opportunity_id DESC
        Sort Method: quicksort  Memory: 25kB
        ->  Nested Loop  (cost=10.30..23.71 rows=4 width=207) (actual time=0.321..0.322 rows=0 loops=1)
              Join Filter: (current_opportunity_summary.opportunity_summary_id = link_opportunity_summary_funding_category.opportunity_summary_id)
              ->  Nested Loop  (cost=10.01..22.65 rows=3 width=239) (actual time=0.321..0.322 rows=0 loops=1)
                    Join Filter: (current_opportunity_summary.opportunity_summary_id = link_opportunity_summary_funding_instrument.opportunity_summary_id)
                    ->  Nested Loop  (cost=9.73..21.94 rows=2 width=231) (actual time=0.321..0.321 rows=0 loops=1)
                          ->  Nested Loop Semi Join  (cost=9.44..21.58 rows=1 width=223) (actual time=0.320..0.321 rows=0 loops=1)
                                Join Filter: (opportunity.opportunity_id = current_opportunity_summary_1.opportunity_id)
                                ->  Nested Loop  (cost=9.16..21.25 rows=1 width=231) (actual time=0.320..0.321 rows=0 loops=1)
                                      ->  Nested Loop  (cost=8.87..20.91 rows=1 width=223) (actual time=0.320..0.321 rows=0 loops=1)
                                            ->  Bitmap Heap Scan on opportunity  (cost=8.59..12.60 rows=1 width=207) (actual time=0.320..0.320 rows=0 loops=1)
"                                                  Recheck Cond: ((agency = 'US-MMM'::text) OR (agency = 'HHS'::text))"
                                                  Filter: (is_draft IS FALSE)
                                                  ->  BitmapOr  (cost=8.59..8.59 rows=1 width=0) (actual time=0.319..0.319 rows=0 loops=1)
                                                        ->  Bitmap Index Scan on opportunity_agency_idx  (cost=0.00..4.29 rows=1 width=0) (actual time=0.301..0.301 rows=0 loops=1)
"                                                              Index Cond: (agency = 'US-MMM'::text)"
                                                        ->  Bitmap Index Scan on opportunity_agency_idx  (cost=0.00..4.29 rows=1 width=0) (actual time=0.017..0.017 rows=0 loops=1)
"                                                              Index Cond: (agency = 'HHS'::text)"
                                            ->  Index Scan using current_opportunity_summary_opportunity_id_idx on current_opportunity_summary  (cost=0.29..8.30 rows=1 width=16) (never executed)
                                                  Index Cond: (opportunity_id = opportunity.opportunity_id)
                                      ->  Index Only Scan using opportunity_summary_pkey on opportunity_summary  (cost=0.29..0.34 rows=1 width=8) (never executed)
                                            Index Cond: (opportunity_summary_id = current_opportunity_summary.opportunity_summary_id)
                                            Heap Fetches: 0
                                ->  Index Only Scan using current_opportunity_summary_opportunity_id_idx on current_opportunity_summary current_opportunity_summary_1  (cost=0.29..0.32 rows=1 width=8) (never executed)
                                      Index Cond: (opportunity_id = current_opportunity_summary.opportunity_id)
                                      Heap Fetches: 0
                          ->  Index Only Scan using link_opportunity_summary_applicant_type_opportunity_sum_fee2 on link_opportunity_summary_applicant_type  (cost=0.29..0.34 rows=2 width=8) (never executed)
                                Index Cond: (opportunity_summary_id = current_opportunity_summary.opportunity_summary_id)
                                Heap Fetches: 0
                    ->  Index Only Scan using link_opportunity_summary_funding_instrument_opportunity_4597 on link_opportunity_summary_funding_instrument  (cost=0.29..0.33 rows=2 width=8) (never executed)
                          Index Cond: (opportunity_summary_id = link_opportunity_summary_applicant_type.opportunity_summary_id)
                          Heap Fetches: 0
              ->  Index Only Scan using link_opportunity_summary_funding_category_opportunity_s_9f79 on link_opportunity_summary_funding_category  (cost=0.29..0.33 rows=2 width=8) (never executed)
                    Index Cond: (opportunity_summary_id = link_opportunity_summary_applicant_type.opportunity_summary_id)
                    Heap Fetches: 0
Planning Time: 25.776 ms
Execution Time: 0.652 ms
```